keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1313 @@
+* A public key, PKIX-serialized.  This key must be an RSA key with modulus size
+  3072 or 4096, an ECDSA key with curve P256 or P384, or an ED25519 key.
+* A proof of possession of the corresponding private key.
+  * If the key is an RSA key, then the signature is over the ASCII bytes of the
+	  pod UID, using RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as implemented
+	  by the golang function crypto/rsa.SignPKCS1v15).
+  * If the key is an ECDSA key, then the signature is as described by [SEC 1,
+	  Version 2.0](https://www.secg.org/sec1-v2.pdf) (as implemented by the golang
+	  library function crypto/ecdsa.SignASN1)
+  * If the key is an ED25519 key, then the signature is as described by the
+    [ED25519 Specification](https://ed25519.cr.yp.to/) (as implemented by
+    the golang library crypto/ed25519.Sign).
+* Details of the requesting Pod's identity:
+  * Namespace (indicated by the namespace the PodCertificateRequest is created in).
+  * Pod name
+  * Pod UID
+  * Service account name
+  * Service account UID
+  * Node name
+  * Node UID
+* The signer name from which a certificate is being requested.
+* (In the status subobject): An issued certificate chain, stored separated into
+  individual X.509 DER-formatted certificates.  If there is more than one issued
+  certificate, the first element of the chain is assumed to be the leaf
+  certificate, followed by any necessary intermediates in leaf-to-root order.
+* (In the status subobject): A timestamp at which time the requester that made
+  the PodCertificateRequest should begin trying to refresh the certificate.
+
+PodCertificateRequest objects have a simplified lifecycle compared to
+CertificateSigningRequests.  There are no separate approval and signing stages.
+They can only be issued (and thus implicitly approved), denied (represented by a "Denied" status), or failed (represented by a "Failed" status).
+
+Similar to CertificateSigningRequests, controllers that want to issue a
+certificate in response to a PodCertificateRequest need to have two separate RBAC permissions:
+* `update` on `podcertificaterequests/status`.
+* `sign` on the requested signer.
+
+Once a signer controller decides that a PodCertificateRequest is addressed to it by matching on the signer name, it should take one of the following actions:
+* Issue the certificate by updating the status subresource of the
+  PodCertificateRequest, setting both the `certificateChain` and
+  `beginRefreshAt` fields.
+* Reject the certificate by updating the status subresource of the
+  PodCertificateRequest, adding a "Denied" condition.
+* Indicate failure in issuing the certificate by updating the status subresource
+  of the PodCertificateRequest, adding a "Failed" condition.
+
+PodCertificateRequest validation logic will:
```

Comment:
Can we be very clear here about what the _kubelet_ does versus what the _signer_ does, and who owns which step.

I like the idea of a PodCertificateRequest generally, but I am a bit concerned with statements like

> * Confirm that the public key is one of the supported key types.
> * Confirm that the proof-of-possession is valid.
> * Confirm that the issued chain (if one is set) consists of valid certificates, that do indeed form a chain to each other.

that we are effectively creating a new CSR-style API that by-design only the `kubelet` can ever use, and granting the kubelet too much responsibility here.

As it sits, unless you are OK with the specifics of how the `kubelet` attests, validates, and mints certs, the API is sort of useless, and your options are

1. Ignore this  entire API
2. Fork the kubelet

Additionally, I am not sure how much work we are really saving for signers here, by creating a new API and putting the kubelet as the lowest-common-denomintor between the API and the signers.
 
I would rather do something than nothing here, but I wonder if we would we be better served by separating the PodCertificateRequest/Status API stuff from the actual kubelet implementation restrictions?

That could be as simple as kubelet feature-flagging that leaves the PodCertificateRequest API in place but disables the kubelet-specific bits, so the signer (or an intermediary controller) could handle these things using the same API.

Reply:
I'm not quite sure I understand this concern, but I might be missing something.

As it stands, the only kubelet-specific part of the API is that kubelet has support for projected volumes that result in the creation of PodCertificateRequest objects.  Additionally, we'll probably check that PodCertificateRequest objects are created by `system:node:*` users.

It's an explicit goal that third-party signers should be able to plug in to this system, without having to reimplement anything.  They just need to provide a controller that listens for PodCertificateRequests that are directed to their signer name.  There should be no need to replace any part of the kubelet or kube-apiserver support.

Reply:
I believe the validation described here is strictly about the apiserver-side validation, not the kubelet. The important distinction is that this kind of validation will happen only during create/update actions.
Below, there's a description of noderestriction admission plugin, which does a consistency check combining the PCR with an actual pod.

Reply:
Yes, Kubelet does no validation.  It just creates the PodCertificateRequest objects.